Persistent stats: periodically persist RocksDB stats

miasantreble opened the pull request:
This PR introduces persistent stats feature which automatically writes RocksDB statistics back into RocksDB periodically. The stats are written to a builtin column family `_persistent_stats` which is maintained without requiring user intervention. User will be able to access the new column family via public API `DBImpl::PersistentStatsColumnFamily()`.

riversand963 commented on db/version_set.h:
```diff
@@ -1023,6 +1026,7 @@ class VersionSet {
   Env* const env_;
   const std::string dbname_;
   const ImmutableDBOptions* const db_options_;
+  MutableDBOptions* mutable_db_options_;
```
We can change `mutable_db_options_` dynamically. What happens if `stats_persist_period_sec` is zero when `Recover` is called, but changed to a non-zero value later? When will the `_persistent_stats` column family be created?

miasantreble replied:
good question, right now in SetOptions we only check if RepeatableThread needs to be restarted. I guess we need to add a helper function to handle the case where the column family is not created

riversand963 replied:
Maybe we should always create this column family. What do you think?

yiwu-arbug commented on db/db_impl.cc:
```diff
@@ -599,7 +613,64 @@ void DBImpl::StartTimedTasks() {
             stats_dump_period_sec * 1000000));
       }
     }
+    stats_persist_period_sec = mutable_db_options_.stats_persist_period_sec;
```
access `mutable_db_options` and (I think) `thread_persist_stats_` should hold mutex.

yiwu-arbug commented on db/db_impl.cc:
```diff
@@ -599,7 +613,64 @@ void DBImpl::StartTimedTasks() {
+  std::string now_micros_string = std::to_string(now_micros);
+  int keycount = 0;
+  for (auto iter = stats_map.begin(); iter != stats_map.end(); ++iter) {
+    // TODO(Zhongyi) use more readable timestamp?
```
I find this: https://github.com/facebook/rocksdb/blob/master/env/posix_logger.h#L111

yiwu-arbug commented on db/db_impl.cc:
```diff
@@ -599,7 +613,64 @@ void DBImpl::StartTimedTasks() {
+  WriteOptions wo;
+  ColumnFamilyOptions cf_options;
+  const uint64_t now_micros = env_->NowMicros();
+  std::string now_micros_string = std::to_string(now_micros);
```
nit: use `ToString` from util/string_util.h

yiwu-arbug commented on db/db_impl.cc:
```diff
@@ -790,6 +861,22 @@ Status DBImpl::SetDBOptions(
+          mutable_db_options_.stats_persist_period_sec) {
+          if (thread_persist_stats_) {
+            mutex_.Unlock();
+            thread_persist_stats_->cancel();
```
Since we have a lot of code with this pattern, how about we extend `RepeatableThread::cancel` with

```
void RepeatableThread::cancel(Mutex* mutex) {
    if (mutex != nullptr) {
        mutex->unlock();
    }
    // actual cancel code
    if (mutex != nullptr) {
        mutex->lock();
    }
}

yiwu-arbug commented on db/db_impl_open.cc:
```diff
@@ -412,6 +412,12 @@ Status DBImpl::Recover(
     default_cf_handle_ = new ColumnFamilyHandleImpl(
         versions_->GetColumnFamilySet()->GetDefault(), this, &mutex_);
     default_cf_internal_stats_ = default_cf_handle_->cfd()->internal_stats();
+    if (mutable_db_options_.stats_persist_period_sec != 0) {
```
Shall we always open the persist stats CF, since you provide an `PersistentStatsColumnFamily` API?

riversand963 replied:
+1. If the user does not want to use it, the column family can still be there with very low overhead.

siying replied:
@yiwu-arbug @riversand963 we can't do that. We need an option to turn it on to keep the DB being able to be downgraded to an older version that can't consume this CF.

riversand963 replied:
I see @siying . Just to confirm. If we start with a new version of RocksDB enabling persistent stats column family, and downgrade to an old version that does not support this feature, the persistent stats column family will be treated as an ordinary column family by the old version?

siying replied:
@riversand963 but since users aren't aware of stats CF (at least in their older version), they will fail to open this DB. This can create a maintenance challenge for them.

yiwu-arbug commented on db/internal_stats.cc:
```diff
@@ -1037,6 +1044,44 @@ void InternalStats::DumpDBStats(std::string* value) {
+  double seconds_up = (env_->NowMicros() - started_at_ + 1) / kMicrosInSec;
+  double interval_seconds_up = seconds_up - db_stats_snapshot_.seconds_up;
+  (*db_stats)["seconds_up"] = std::to_string(seconds_up);
+  (*db_stats)["interval_seconds_up"] = std::to_string(interval_seconds_up);
```
nit: use `ToString` from util/string_util.h

yiwu-arbug commented on db/internal_stats.cc:
```diff
@@ -1037,6 +1044,44 @@ void InternalStats::DumpDBStats(std::string* value) {
+  (*db_stats)["write_stall_micros_readable"] = std::string(human_micros);
+
+  // Interval
+  (*db_stats)["interval_write_other"] = std::to_string(GetDBStats(InternalStats::WRITE_DONE_BY_OTHER) - db_stats_snapshot_.write_other);
```
nit: seems line too long. run `make format` once?

siying commented on db/db_impl.cc:
```diff
@@ -97,6 +97,7 @@
 
 namespace rocksdb {
 const std::string kDefaultColumnFamilyName("default");
+const std::string kPersistentStatsColumnFamilyName("_persistent_stats");
```
@yiwu-arbug is correct. This is a case we should absolutely handle. It can be as simple as skipping this feature if we have a conflict name.

Also, "_persistent_stats" is too easy to conflict. We need to add more special characters to make it harder to conflict. "Persistent" doesn't need to appear in the CF name. Whatever written to the DB is persistent.

siying commented on db/db_impl.cc:
```diff
@@ -599,7 +613,64 @@ void DBImpl::StartTimedTasks() {
+  }
+  TEST_SYNC_POINT("DBImpl::PersistStats:2");
+  ROCKS_LOG_WARN(immutable_db_options_.info_log,
+                 "------- PERSISTING STATS -------");
```
Why add a log line?

siying commented on db/db_impl.cc:
```diff
@@ -599,7 +613,64 @@ void DBImpl::StartTimedTasks() {
+  WriteOptions wo;
+  ColumnFamilyOptions cf_options;
+  const uint64_t now_micros = env_->NowMicros();
+  std::string now_micros_string = std::to_string(now_micros);
```
ToString() is not sorted by time order. We should have timestamp sort in time order.

ajkr replied:
`std::to_string` is also not sorted, right? I think we need a fixed-width representation

siying commented on db/db_impl_readonly.cc:
```diff
@@ -137,13 +137,25 @@ Status DB::OpenForReadOnly(const Options& options, const std::string& dbname,
   if (s.ok()) {
-    assert(handles.size() == 1);
+    if (db_options.stats_persist_period_sec != 0) {
+      assert(handles.size() == 2);
```
The interface assumes that the stat CF is a normal CF that users should access using Get() API. Though this may not be a bad idea, this heavily violate backward compatibility. We should think twice about changes breaking backward compatibility.

riversand963 replied:
I think the stat CF is very similar to a normal CF. Do you mean we should not expose it via `handles`? @siying

ajkr replied:
I'd prefer an API that returns timeseries stats in a structured way and hides the separate CF. Exposing the raw key-values, where keys are encoded in a specific way, feels a bit low-level.

riversand963 replied:
If we return time series stats in a structured way while the underlying storage is kv storage, I think we are somewhat trying to define and implement a 'data model' on top of persistent stats storage. My suggestion would be keeping it simple and clear. This also reminds me of MySQL's `performance_schema`. The data model of `performance_schema` is SQL because MySQL supports SQL. If RocksDB supports only kv storage, then what kind of data model should we support in this case? The answer to this question also largely depends on what kind of queries we would like RocksDB to be able to answer about performance stats. I think supporting the data retrieval within a given time range should be supported. Besides that, do we want to support aggregation? I don't have a clear answer to this, just post the question so that we can take it into account when we discuss.

siying commented on db/version_set.cc:
```diff
@@ -3240,11 +3242,19 @@ Status VersionSet::ApplyOneVersionEdit(
-    if (cf_options == name_to_options.end()) {
+    // implicitly add persistent_stats column family without requiring user
+    // to specify
+    bool isPersistStatsColumnFamily = edit.column_family_name_.compare(
```
Naming convention is `is_persistent_stats_column_family`.

siying commented on db/version_set.cc:
```diff
@@ -3240,11 +3242,19 @@ Status VersionSet::ApplyOneVersionEdit(
     } else {
-      cfd = CreateColumnFamily(cf_options->second, &edit);
+      if (isPersistStatsColumnFamily) {
+        cfd = CreateColumnFamily(ColumnFamilyOptions(), &edit);
```
Creating this column family options in this way may not be good for stats. For example, it uses a 64MB memtable, and create a new 8MB block cache. We should be careful about the resource this CF uses.

miasantreble replied:
since we are hiding persistent stats column family from users, user will not be able to specify CFOptions for this CF, maybe we should have a LITE version of CFOptions?

siying commented on db/version_set.cc:
```diff
@@ -3471,6 +3487,25 @@ Status VersionSet::Recover(
+      BaseReferencedVersionBuilder* builder =
+          new BaseReferencedVersionBuilder(persistent_stats_cfd);
+      builder->version_builder()->Apply(&persistent_stats_cf_edit);
+      builders.insert({next_id, builder});
```
Can we reuse existing CreateColumnFamily() code? Actually do we need to do it in VersionSet? Can we do this in DBImpl after VersionSet::Recover() returns?

riversand963 replied:
Reusing `DBImpl::CreateColumnFamilyImpl` makes sense because we support dynamic changing of `stats_persist_period_sec` and we also support dynamic adding/dropping column families.
Therefore, `DBImpl::Recover` can check whether `stats_persist_period_sec` is non-zero AND the `stats` CF has been created during `VersionSet::Recover`. If not, create the column family. We also need an API. When the user changes the `stats_persist_period_sec` from 0 to non-zero, we call `DBImpl::CreateColumnFamilyImpl`.

siying commented on db/version_set.cc:
```diff
@@ -2715,6 +2716,7 @@ VersionSet::VersionSet(const std::string& dbname,
       env_(_db_options->env),
       dbname_(dbname),
       db_options_(_db_options),
+      mutable_db_options_(mutable_db_options),
```
If it is only used in Recover(), it should be passed as an option in Recover().

siying commented on include/rocksdb/options.h:
```diff
@@ -645,6 +645,10 @@ struct DBOptions {
   // Default: 600 (10 min)
   unsigned int stats_dump_period_sec = 600;
 
+  // if not zero, dump rocksdb.stats to LOG every stats_persist_period_sec
```
To LOG?

miasantreble commented:
@ajkr I was thinking about adding an ldb command to read the persistent_stats column family and nicely print the stats

ajkr commented:
Oh ok. I hope we consider returning stats in a structured form so people (including us) can query them programmatically without having to parse the result. Plus there's the benefit of not exposing low-level details to users: we can change the implementation details later if we want.

riversand963 commented on db/db_impl_open.cc:
```diff
@@ -412,6 +432,16 @@ Status DBImpl::Recover(
+    // the CF is already created in VersionSet::ApplyOneVersionEdit, but
+    // column family handle was not. Need to explicitly create handle here.
+    if (mutable_db_options_.stats_persist_period_sec != 0 &&
+        persist_stats_cf_handle_ == nullptr) {
```
If we reach here, is it possible that `mutable_db_options_.stats_persist_period_sec != 0 && persist_stats_cf_handle_ != nullptr` ?

miasantreble replied:
Yes, when we are recovering from a RocksDB instance that perviously had persistent_stats column family created, we would need to skip the `CreateColumnFamily` call above as `ApplyOneVersionEdit` would have already created the column family. However, the column family handle still needs to be explicitly created here

riversand963 commented on db/db_impl_open.cc:
```diff
@@ -412,6 +432,16 @@ Status DBImpl::Recover(
+              kPersistentStatsColumnFamilyName),
+          this, &mutex_);
+    }
     single_column_family_mode_ =
```
`single_column_family_mode_ = versions_->GetColumnFamilySet()->NumberOfColumnFamilies() == 1;` is no longer true if there is persistent stats column family.

miasantreble replied:
yes, this needs to be addressed

riversand963 commented on db/db_impl_open.cc:
```diff
@@ -1044,12 +1074,23 @@ Status DB::Open(const Options& options, const std::string& dbname, DB** dbptr) {
+    }
     // i can delete the handle since DBImpl is always holding a reference to
     // default column family
+    if (db_options.stats_persist_period_sec != 0) {
```
Maybe `if (handles[1] != nullptr)`

miasantreble commented on db/db_impl_open.cc:
```diff
@@ -386,11 +386,31 @@ Status DBImpl::Recover(
   }
   if (s.ok() && !read_only) {
     for (auto cfd : *versions_->GetColumnFamilySet()) {
+      if (cfd->GetName().compare(kPersistentStatsColumnFamilyName) == 0) {
```
Good catch, I think you are right, we should only skip this step when persistent_stats CF is newly created

miasantreble commented:
summary of recent commits (72409f0 ~ 19e4d96): 
* Added `GetStatsHistory` API to get stats history by giving a time range specified in `GetStatsOptions` from a live DB. `GetStatsHistory` will return a map from timestamp to stats history (another map from stats name to stats value). For example, time_stamp = 1543432534530431, stats_key = compaction.L0.AvgSec, stats_value = 0.001359.
* Separated `persist_stats_to_disk` from `stats_persist_period_sec` to control whether to write stats to RocksDB (as opposed to merely storing in memory), previously `stats_persist_period_sec` = 0 means skip persisting to disk. This should allow users to use `GetStatsHistory` to access stats history since the latest DB restart without worrying about a new column family being created implicitly and increased downgrade operational costs. 
* Added unit test to verify when `persist_stats_to_disk` is enabled, `GetStatsHistory` will return both in memory stats history and stats history recovered from disk (before the latest DB restart)
@siying PTAL

siying commented:
Can you rebase?

siying commented:
Never mind. I find https://github.com/facebook/rocksdb/pull/5046 . So maybe we should close this one.
